packages/liveblocks-client/src/shallow.test.ts

```diff
@@ -0,0 +1,150 @@
+import fc from "fast-check";
+
+import shallow from "./shallow";
+
+const scalar = () => fc.jsonValue({ maxDepth: 0 });
+
+const complex = () =>
+  fc.anything().filter((value) => value !== null && typeof value === "object");
+
+describe("shallow", () => {
+  it("scalar values", () => {
+    expect(shallow(0, 0)).toBe(true);
+    expect(shallow("", "")).toBe(true);
+    expect(shallow("hi", "hi")).toBe(true);
+    expect(shallow(false, false)).toBe(true);
+    expect(shallow(true, true)).toBe(true);
+
+    expect(shallow(0, 1)).toBe(false);
+    expect(shallow(false, true)).toBe(false);
+    expect(shallow(true, false)).toBe(false);
+
+    // Weird exceptions
+    expect(shallow(NaN, NaN)).toBe(true);
+    expect(shallow(-0, +0)).toBe(false);
+  });
+
+  it("scalar values wrapped in list", () => {
+    expect(shallow([0], [0])).toBe(true);
+    expect(shallow([""], [""])).toBe(true);
+    expect(shallow(["hi"], ["hi"])).toBe(true);
+    expect(shallow([false], [false])).toBe(true);
+    expect(shallow([true], [true])).toBe(true);
+
+    expect(shallow([0], [1])).toBe(false);
+    expect(shallow([false], [true])).toBe(false);
+    expect(shallow([true], [false])).toBe(false);
+
+    // Weird exceptions   ],[
+    expect(shallow([NaN], [NaN])).toBe(true);
+    expect(shallow([-0], [+0])).toBe(false);
+  });
+
+  it("scalar values wrapped in objs", () => {
+    expect(shallow({ k: 0 }, { k: 0 })).toBe(true);
+    expect(shallow({ k: "" }, { k: "" })).toBe(true);
+    expect(shallow({ k: "hi" }, { k: "hi" })).toBe(true);
+    expect(shallow({ k: false }, { k: false })).toBe(true);
+    expect(shallow({ k: true }, { k: true })).toBe(true);
+
+    expect(shallow({ k: 0 }, { k: 1 })).toBe(false);
+    expect(shallow({ k: false }, { k: true })).toBe(false);
+    expect(shallow({ k: true }, { k: false })).toBe(false);
+
+    // Weird exceptions
+    expect(shallow({ k: NaN }, { k: NaN })).toBe(true);
+    expect(shallow({ k: -0 }, { k: +0 })).toBe(false);
+  });
+});
+
+describe("shallow (properties)", () => {
+  it("all equal scalars are shallowly equal", () => {
+    fc.assert(
+      fc.property(
+        // Inputs
+        fc.clone(scalar(), 2),
+
+        // Unit test
+        ([scalar1, scalar2]) => {
+          expect(shallow(scalar1, scalar2)).toBe(true);
+
+          // If two values are shallowly equal, then those values wrapped in an
+          // array (one level) should _also_ be considered shallowly equal
+          expect(shallow([scalar1], [scalar2])).toBe(true);
+          expect(shallow([scalar1, scalar1], [scalar2, scalar2])).toBe(true);
+
+          // Ditto for objects
+          expect(shallow({ a: scalar1 }, { a: scalar2 })).toBe(true);
+          expect(
+            shallow({ a: scalar1, b: scalar1 }, { a: scalar2, b: scalar2 })
+          ).toBe(true);
+        }
+      )
+    );
```

Comment:
Here is the **tl;dr** for a typical property-based test with the `fast-check` library:

The tests are always structured like so:

```
fc.assert(
  fc.property(
    // Input generators
    fc.float(),  // "x"
    fc.float(),  // "y"
    ...

    // Unit test
    (x, y, ...) => {
      // "Normal" jest expects
      expect(x + y).toBe(y + x);   // Example
    }
  )
)
```

First, `fc.assert()` will drive the test, by running it many times. It generates random input values, runs the tests, and if the tests fail, it will "shrink" the failing input to a smaller counter example that also fails. Then the counter example gets reported in a failing test.